src/crypto/aes.js

```diff
@@ -0,0 +1,89 @@
+import CID from '../cid.js'
+import random from 'js-crypto-random'
+import aes from 'js-crypto-aes'
+
+/**
+ * @param {number} value
+ */
+const enc32 = value => {
+  value = +value
+  const buff = new Uint8Array(4)
+  buff[3] = (value >>> 24)
+  buff[2] = (value >>> 16)
+  buff[1] = (value >>> 8)
+  buff[0] = (value & 0xff)
+  return buff
+}
+
+/**
+ * @param {Uint8Array} buffer
+ */
+const readUInt32LE = (buffer) => {
+  const offset = buffer.byteLength - 4
+  return ((buffer[offset]) |
+          (buffer[offset + 1] << 8) |
+          (buffer[offset + 2] << 16)) +
+          (buffer[offset + 3] * 0x1000000)
+}
+
+/**
+ * @param {Uint8Array[]} buffers
+ */
+const concat = buffers => Uint8Array.from(buffers.map(b => [...b]).flat())
+
+/**
+ * @template {'aes-gcm' | 'aes-cbc' | 'aes-ctr'} Name
+ * @template {number} Code
+ * @param {Object} options
+ * @param {Name} options.name
+ * @param {Code} options.code
+ * @param {number} options.ivsize
+ */
+const mkcrypto = ({ name, code, ivsize }) => {
+  // Line below does a type cast, because type checker can't infer that
+  // `toUpperCase` will result in desired string literal.
+  const cyperType = /** @type {import('js-crypto-aes/dist/params').cipherTypes} */(name.toUpperCase())
+  /**
+   * @param {Object} options
+   * @param {Uint8Array} options.key
+   * @param {Object} options.value
+   * @param {Uint8Array} options.value.bytes
+   * @param {Uint8Array} options.value.iv
+   */
+  const decrypt = async ({ key, value: { iv, bytes } }) => {
+    bytes = await aes.decrypt(bytes, key, { name: cyperType, iv, tagLength: 16 })
+    const len = readUInt32LE(bytes.subarray(0, 4))
+    const cid = CID.decode(bytes.subarray(4, 4 + len))
+    bytes = bytes.subarray(4 + len)
+    return { cid, bytes }
+  }
+  /**
+   * @param {Object} options
+   * @param {Uint8Array} options.key
+   * @param {Uint8Array} options.bytes
+   * @param {CID} options.cid
+   */
+  const encrypt = async ({ key, cid, bytes }) => {
```

Comment:
I think this is yet another time where I wish CID+block bytes had a proper name to go by.

Reply:
I actually find this to be a much better pattern because there are all kinds of ways that you may arrive at a CID + Bytes (car files reads, blocks, encrypted block decodes, etc) and destructuring them into only the parts these functions need is a great practice that gives the function maximal utility. Once you “name” it you end up with a type, and with blocks that tends to mean that you also have that codec and a decoded state, which is only useful when it is and is a burden when it’s not needed.